### ISSUE TITLE
Can TGI load an 8bit or 4bit model?

### ISSUE THREAD

**Issue by RonanKMcGovern** [huggingface/text-generation-inference#1410]
Bitsandbytes now allows the pushing of 4bit models to hub (this was already possible for 8bit models). Bitsandbytes supports 4bit, nf4 and 8bit formats. Any of these models should correctly load using model.AutoModelForCausalLM (see [docs](https://huggingface.co/docs/transformers/main/quantization?bnb=4-bit)).

## Will TGI correctly load a model that is saved on the hub as nf4?
I assume yes?

## Can TGI load and inference a model that is 8bit on the hub?
Since --quantize bitsandbytes (which was for 8bit models) is deprecated, it is automatically replaced by eetq. However, eetq isn't necessarily compatible with the 8bit quantization of bitsandbytes.

Therefore, is there a way to inference a model loaded from 8bits?

Tagging @Narsil

**Comment by Narsil** [comment]
>  it is automatically replaced by eetq. 

No it doesn't, it simply tells about the deprecation. bnb-nf4 and bnb-fp4 are functional (just use those flags).

EETQ just happens to work just as well as bnb (8bits) and is much faster on the inference latency.

**Comment by RonanKMcGovern** [comment]
**Regarding 4bit**
Thanks @Narsil , I guess the issue is that the TGI docker image isn't using the latest dev version of transformers, which is required AFAIK to load 4-bit models, see the note from poedator [here](https://github.com/TimDettmers/bitsandbytes/pull/753#issuecomment-1881121385). I suppose I could just manually make a tweaked docker image?

**Regarding 8bit**
So the 8bit quant method used by EETQ must be the same (or sufficiently similar) to the bnb 8bit method then? (if EETQ is to just work out of the box when an 8bit model is loded)

**Comment by Narsil** [comment]
> which is required AFAIK to load 4-bit models, see the note from poedator https://github.com/TimDettmers/bitsandbytes/pull/753#issuecomment-1881121385. I suppose I could just manually make a tweaked docker image?

TGI doesn't use transformers at all (It's a fallback for unsupported models, so might fail).
You don't need the saved 4bits models to actually use them, the quantization will happen on the fly (that's the main interest of BNB, otherwise I would suggest looking at AWQ, which is much faster for inference, GPTQ being another alternative).

**Comment by RonanKMcGovern** [comment]
Yeah, I understand how the on-the-fly quanting works. However, the use-case here is to save on downloading time by directly loading the model in quantized form (whether nf4 or 8bit).

Yes, it's possible to directly load AWQ (better accuracy and speed than GPTQ) but that requires quantization in advance, which takes compute and time after having done a fine-tuning (and AWQ adapters cannot be merged to the base model).

Being able to directly download and inference a 4 bit or 8 bit model into TGI is therefore quite useful.

I have done some initial testing on feeding an 8bit model into TGI and the model is responding with blanks, but I'm going to set up a clear test and then I'll post here with an exact replicator of that.

I could try out an nf4 model as well, but I'm guessing that isn't supported then yet given the serialization of 4bit models is quite new?

**Comment by RonanKMcGovern** [comment]
Probably this should just be a feature request then.

Now that transformers allows 4bit (and 8bit) to be pushed to the hub, it would be attractive to inference those models right away with TGI.

The shortest current pathway from fine-tuning to inference is:
- Train with transformers and save in bf16.
- Quantize to AWQ and push to hub in 4bit.
- Inference with AWQ in TGI.

OR 
- Train with transformers and save+push in bf16.
- Inference with eetq in TGI.

This is good, but model weights are big.

~
If TGI could inference nf4 or (especially) 8bit models, that would allow:
- fine-tune with transformers and save+push in 8bit.
- Inference right away with TGI.

OR 
- fine-tune with transformers and save+push in nf4
- Inference right away with TGI.

which will be slower to inference, but a quick way to get from fine-tuning to inference with small weights.

**Comment by Narsil** [comment]
Then sorry but not we're most likely not going to implement that.

BNB is really great to quantize models which are not already quantized, but the inference performance is just not there (compared to like AWQ), therefore unless there's a more compelling reasons, we're not going to spend time adding new surface and maintenance to load directly the prequantized weights. It just doesn't seem it brings enough value (cut down a little on network and disk space, which is infinitely cheaper than GPUs).
Spawn times are important, but since there's another way (using AWQ) that just seems strictly superior from an inference point of view I don't think we want to go down that road.

**Comment by RonanKMcGovern** [comment]
Understandable, appreciate the response.

**Comment by nolwennz** [comment]
Hello, 

We quantized Mixtral 8x7b Instruct using bitsandbytes nf4 locally (with some specific parameters and excluding some weights files) and we have great performances we don't meet with the --quantize bitsandbytes-nf4 flag. 
We tried to run our 4-bit model with TGI but we get an assertion error on the weight.shape list. 

From what I understand from the thread, it is not possible to run models with prequantized weights (except for those with AWQ and GPTQ)? 
Alternatively, is there a way to pass some custom parameters for bitsandbytes-nf4 quantization? (bnb_4bit_use_double_quant=True, bnb_4bit_compute_dtype=torch.bfloat16)

Many thanks :pray:

**Comment by farzanehnakhaee70** [comment]
Isn't ther still no possibility to use the models with prequantized weights?
I want to use `unsloth/phi-4-bnb-4bit` model. When I deploy in python with transformer libraries, it only occupies 8GB of memoey. However, in TGI, it raises OOM which I think it changed it to float32 again.
